internal/impl/snowflake/streaming/parquet.go

```diff
@@ -11,76 +11,86 @@
 package streaming
 
 import (
+	"bytes"
 	"encoding/binary"
 	"fmt"
-	"io"
 
 	"github.com/parquet-go/parquet-go"
 	"github.com/parquet-go/parquet-go/format"
 	"github.com/redpanda-data/benthos/v4/public/service"
 	"github.com/segmentio/encoding/thrift"
 )
 
-func messageToRow(msg *service.Message) (map[string]any, error) {
+// messageToRow converts a message into columnar form using the provided name to index mapping.
+// We have to materialize the column into a row so that we can know if a column is null - the
+// msg can be sparse, but the row must not be sparse.
+func messageToRow(msg *service.Message, out []any, nameToPosition map[string]int) error {
 	v, err := msg.AsStructured()
 	if err != nil {
-		return nil, fmt.Errorf("error extracting object from message: %w", err)
+		return fmt.Errorf("error extracting object from message: %w", err)
 	}
 	row, ok := v.(map[string]any)
 	if !ok {
-		return nil, fmt.Errorf("expected object, got: %T", v)
+		return fmt.Errorf("expected object, got: %T", v)
 	}
-	mapped := make(map[string]any, len(row))
 	for k, v := range row {
-		mapped[normalizeColumnName(k)] = v
+		idx, ok := nameToPosition[normalizeColumnName(k)]
```

Comment:
this is cool. wondeirng if normalizing to integer and doing integer comparison of a column name would be faster. 

`k=key in the row` would "hash" or whatev to a memoized column name. but probably not where the perf is.

Reply:
`normalizeColumnName` is fairly expensive. I'm thinking about writing a microbench for it and optimizing it, but I think the greater perf at this point would be parallelizing zstd compression in the parquet writing bit. This conversion is already parallelized.

Reply:
Doubled the common case for `normalizeColumnName` here: https://github.com/redpanda-data/connect/pull/2979